[AOP-2285] Improved docs about applying OS patches (#70)
mulesoft/docs-private-cloud: modules/ROOT/pages/restarting-a-node.adoc

= Update Nodes
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> 

Reviewer: No text between the topic set title and prereqs? 
I just made a comment about this in another edit for you today. If you can, try to add something about the circumstances under which your audience wants or needs to perform this task, and any special requirements that the audience should know before starting.
If there is nothing else, I suggest moving the prerequisite text here: something like "After you set up and test the backup procedure on all clusters, or after you back up nodes, you are ready to make cluster and application updates."

== Before You Begin

* Set up and test the backup procedure on all clusters.
* Perform a backup before attempting node maintenance to preserve data in the event of a failure.

== Cluster and Application Updates

> [IMPORTANT]

Reviewer: The topic title should be followed by text that identifies these types of updates and the circumstance under which the audience wants or needs to make them.
The topic set intro should not comprise one IMPORTANT note.
As above, if there is nothing to say about the updates and why, then the prereqs become the intro, and should be worded as things already done rather than to be done.

> The following steps must be performed on one Anypoint Platform Private Cloud Edition node at a time to minimize or eliminate component breakage during the update process. The entire process must be completed on each node before the next node is addressed. Verify the health status of all cluster components and applications after each node update before proceeding with the next node.

Reviewer: ```suggestion
You must perform cluster and application updates on one Anypoint Platform Private Cloud Edition node at a time to minimize or eliminate component breakage during the update process. You must complete the entire process and verify the health of all cluster components on one node before proceeding to the next node.
```

Reviewer: BUT...this intro text names user actions in a different order than the following three tasks are presented.
Please make the order in the intro and the three subtopics match.

Reviewer: In fact, consider simplifying the intro text as follows:
```suggestion
To make cluster and application updates, you must perform the following tasks completely and in order, one node at a time:
. Verify the status of cluster health.
. Install updates.
. Verify installation success.
```

=== Verify Cluster Health Status

> Verify that the cluster is healthy before starting the process. Otherwise, many of the assumptions made in this process are not valid.

Reviewer: More to the point:

"If your cluster is not healthy before you begin the upgrade, the upgrade fails. You test the health of your cluster by verifying certain command output and verifying that all pods are running."

> . Verify the output from the following commands:

Reviewer: ```suggestion
. Verify status and cluster health:
```

+
```
gravity status
etcdctl cluster-health
```
+

> If the output from the previous commands mentions a `degraded` state or `errors`, fix all issues before proceeding.

Reviewer: ```suggestion
If the output mentions a `degraded` state or `errors`, fix all issues before proceeding.
```
Can we point them to answers or troubleshooting?

. Verify that all pods are running:
+
```
gravity get pod -o wide--all-namespaces | grep -vi Running
```
+

> If the output from the previous command lists any pod not in "Running" state, then fix all issues before proceeding.

Reviewer: ditto edits to followup sentence to peer list item above

=== Install Updates

> After verifying that the cluster is healthy, select a node on which to perform the following steps:

Reviewer: This intro is not really helpful, because it tells the audience to select a node but gives no criteria other than that they are going to perform installation on it.
I think we might want something more like "After you verify the health of the cluster, pick any node on which to install updates. Remember that you must not only complete this installation but also verify it on this one node before you proceed to update and verify on the next node."

> . Retrieve the *name* of the target node:

Reviewer: ```suggestion
. Retrieve the name of the target node:
```

+
```
kubectl get nodes
```
. Drain the node:
+
```
kubectl drain --delete-local-data --ignore-daemonsets <node_name>
```
. After the node is drained, identify the gravity unit name:
+
```
sudo systemctl list-units --plain | awk '/planet-master/{print $1}'
```

> The previous command outputs a `systemd` unit name, which is used as the `$unit_name` value in the next step.

Reviewer: ```suggestion
Take note of the `systemd` unit name, which is used next as the `$unit_name` value.
```

. Stop the gravity unit:
+
```
systemctl stop $unit_name
systemctl status $unit_name
```

> . After the `status` subcommand indicates the gravity unit is stopped, install the needed updates and reboot the server.

Reviewer: ```suggestion
. After the `status` command indicates that the gravity unit is stopped, install the updates and reboot the server.
```
Hmm...no text at all about what the updates are or HOW they are installed? When I see the task "Install <Something>," I expect the actual installation steps, not the steps that prepare for installation.

> === Perform Post-update Check

Reviewer: Verify the Installation

Reviewer: Try to add some snippet of text here that provides the context and leads to the first step of this third task.

. Verify that all systems correctly restarted:
+
```
systemctl status --failed
```

> The previous command should return zero units, which means that your system boot phase was successful and all units started. If any units failed to start, fix all issues before proceeding.

Reviewer: ```suggestion
This command should return zero units, which means that your system reboot is successful and all units started. If the command output is not zero units, fix all issues before proceeding.
```
What is the actual return: 0? 0 units? zero units? 
Also, any info to user re. "fix all issues"? What might those issues be, and how are they fixed? ref to other info?

> . Check the gravity `systemd` units:

Reviewer: Check for what? What type of output is expected? What should the user do about that output?

+
```
systemctl list-units '*gravity__gravitational*'
```

> . Check for errors in the logs by running `journalctl`.

Reviewer: What logs? Is this the full syntax, input from the same CLI as the previous? What would an error look like, and what should the user do?

> . If everything started correctly, run `gravity enter` and verify the results.

Reviewer: "everything started correctly" referring to the previous steps? This goes without saying, if this is the next step.
So, this step should be... what? "Run `gravity enter` and verify the results?" Verify them for what? What should the user see or not see as results?

. Check the cluster status again:
+
```
gravity status
etcdctl cluster-health
gravity get pod -o wide--all-namespaces | grep -vi Running
```

> === Rejoin the Node

Reviewer: This task should be introduced above, along with its peers.
There should also be a transition sentence here that places this task in the context of the process.

. Check the gravity Planet master and teleport status:
+
```
systemctl list-units '*gravity__gravitational*'
```

. If the status of the gravity service is not `loaded`, `active`, or `running`, then you must enable it:
+
```
systemctl enable --now <name_of_the_gravity_unit>
```

> . Uncordon the node:

Reviewer: ```suggestion
. Uncordon the node, so that the target node can resume scheduling pods:
```

+
```
kubectl uncordon <node_name>
```

> The target node is set to again schedule pods. This can take a few minutes to complete.

Reviewer: Suggested moving this definition of "uncordon" to the step action itself. That leaves...:
```suggestion
This can take a few minutes to complete.
```

. Verify cluster status:
+
```
gravity status
```
+

> This command should display a `healthy` status for all nodes of the cluster. After you verify that the node has rejoined the cluster and is in healthy status, repeat the process for the next node in the cluster until all nodes are updated.

Reviewer: ```suggestion
Command output displays a `healthy` status for all nodes of the cluster. Repeat this process for the next node in the cluster until all nodes are updated.
```

include::_attributes.adoc[]
endif::[]

== Before You Begin

* Set up and test the backup procedure on all clusters after you complete installation.
* Perform a backup before attempting node maintenance to preserve data in the event of a failure.

== Cluster and Application Updates

> The following steps must be performed on one Anypoint Platform Private Cloud Edition node at a time to minimize or eliminate component breakage during the update process. The entire process must be completed on each node before the next node is addressed. Verify the health status of all cluster components and applications after each node update before proceeding with the next node.

Reviewer 2: Is there any way to emphasize this?

Reviewer 2: If not, it's ok. but this is really important.

Reviewer 3: Tagged it as important.

Author: is this okay now? not sure I got what's the request...

> This approach assumes that multiple masters are running.

Reviewer 2: we can remove this past of the `masters` as it doesn't add yo our customers.

Author: I didn't write the new version. So I'll leave it as it is as I guess this request is not for me.

Author: @paigehargrave that comment from @federicoboerr is probably for you.

Reviewer 3: I removed the line - thanks!

=== Verify Cluster Status

Verify that the cluster is healthy before starting the process. Otherwise, many of the assumptions made in this process are not valid.

. Verify the output from the following commands:
+
```
gravity status
etcdctl cluster-health
gravity get pod -o wide--all-namespaces | grep -vi Running
gravity get deployment --all-namespaces
gravity get daemonset --all-namespaces
```

> . Review log entries or status message that indicate the possibility of an unhealthy node.

Reviewer 2: @eldios how are customers suppose to run the validations?
Is this step really necessary if they are already following the process?
Customers don't have access to the runbook...
what does it mean for them to `Verify that replication works and is healthy in any replicated software (like Stolon and Cassandra).`?

Author: The point is that if Pods are running but, as an example, Stolon is shouting errors in the logs, then it's not really a green light situation.

In my opinion is important to tell customers to be careful about all components, but it's up to you nice people to decide what your customers should and should not know.

I prefer to be on the safe/cautious side.

The following steps must be performed on one Anypoint Platform Private Cloud Edition node at a time to minimize or eliminate component breakage during the update process. The entire process must be completed on each node before the next node is addressed. Verify the health status of all cluster components and applications after each node update before proceeding with the next node.

=== Verify Cluster Status

Verify that the cluster is healthy before starting the process. Otherwise, many of the assumptions made in this process are not valid.

. Verify the output from the following commands:
+
```
gravity status
etcdctl cluster-health
```

If the output from any of the two commands above mentions `degraded` state or `errors` please be sure to fix the mentioned issue, before proceeding.

Please also verify that all Pods are running by running the commands below:
```
gravity get pod -o wide--all-namespaces | grep -vi Running
```
And if the command above lists any Pod, which will not be in "Running" state, then it means that it's probably having some issues and needs to be fixed.

. Verify that replication works and is healthy in any replicated software (like Stolon and Cassandra).

> -- TODO link to Stolon and Cassandra troubleshooting here

Reviewer 2: @eldios customers don't have access to the troubleshooting documentation for cassandra and stolon.
If we can't provide correct commands, we might as well leave this out.

Author: Removed. Debugging Stolon/Cassandra is a delicate matter and if they don't have access to the full KB then it's better not to deal with that.

> . Check if the firewalld service is running and enabled:

Reviewer 2: @eldios are these steps for `firewalld` necessary? We already cover that during pre-install.
If not, please remove.
